core/gui/src/app/workspace/component/code-editor-dialog/code-editor.component.ts

```diff
@@ -221,6 +224,103 @@ export class CodeEditorComponent implements AfterViewInit, SafeStyle, OnDestroy
+                    return;
+                  }
+
+                  let offset = 0;
+                  let lastLine: number | undefined;
+
+                  this.isMultipleVariables = true;
+                  this.userResponseSubject = new Subject<void>();
+
+                  const processNextVariable = (index: number) => {
+                    if (index >= variablesWithoutAnnotations.length) {
+                      this.isMultipleVariables = false;
+                      this.userResponseSubject = undefined;
+                      return;
+                    }
+
+                    const currVariable = variablesWithoutAnnotations[index];
+
+                    const variableCode = currVariable.name;
+                    const variableLineNumber = currVariable.startLine;
+
+                    // Update range
+                    if (lastLine !== undefined && lastLine === variableLineNumber) {
+                      offset += this.currentSuggestion.length;
+                    } else {
+                      offset = 0;
+                    }
+
+                    const variableRange = new monaco.Range(
+                      currVariable.startLine,
+                      currVariable.startColumn + offset,
+                      currVariable.endLine,
+                      currVariable.endColumn + offset
+                    );
+
+                    const highlight = editor.createDecorationsCollection([
+                      {
+                        range: variableRange,
+                        options: {
+                          hoverMessage: { value: "Argument without Annotation" },
+                          isWholeLine: false,
+                          className: "annotation-highlight",
+                        },
+                      },
+                    ]);
+
+                    this.handleTypeAnnotation(
+                      variableCode,
+                      variableRange,
+                      ed as monaco.editor.IStandaloneCodeEditor,
+                      variableLineNumber,
+                      allCode
+                    );
+
+                    lastLine = variableLineNumber;
+
+                    // Make sure the currVariable will not go to the next one until the user click the accept/decline button
+                    if (this.userResponseSubject !== undefined) {
+                      const userResponseSubject = this.userResponseSubject;
+                      // Only take one response (accept/decline)
+                      const subscription = userResponseSubject
+                        .pipe(take(1))
+                        .pipe(takeUntil(this.workflowVersionStreamSubject))
+                        .subscribe(() => {
+                          highlight.clear();
+                          subscription.unsubscribe();
+                          processNextVariable(index + 1);
+                        });
+                    }
+                  };
```

Comment:
I am not sure if I followed the logic here. a few questions:
1. what's the purpose of `userResponseSubject`? the user should just click a button and that is supposed to accept/decline the annotation suggestion. you should bind a success function to the accept button and a failure function to the decline button. I don't see the need for such a response subject.
2. why it takeUntil `this.workflowVersionStreamSubject`? what's the relationship with workflow version stream?
3. The main logic should prompt the pop up component, for each annotation suggestion. I do not see this logic?

You could explain the logic first. This comment does not intend for you to change the logic. It could just be my understanding issue.

Reply:
1. considering that we may add more user interaction in the future, using a `userResponseSubject` maybe will be more convenient. (use `userResponseSubject` to determine if the user click the button instead of binding function for every actions)
2. I add a `componentDestroy`. It just need a `takeuntil(something)` to avoid the subscribe error. I am not really familiar with this part, so I just copy the code from Xinyuan which is `takeuntil(this.workflowVersionStreamSubject)` before.
3. Yes, the logic is prompting the pop up component for each argument. I use a recursive function `processNextVariable` to iterate the list `variablesWithoutAnnotations` and start from index 0. The highlight and the `handleTypeAnnotation` are added in this recursive function to do operation on each argument individually.